### PR TITLE
[SPARK-17455][MLlib] Improve PAVA implementation in IsotonicRegression

### PR DESCRIPTION
**Pull request by neggert**
## What changes were proposed in this pull request?

New implementation of the Pool Adjacent Violators Algorithm (PAVA) in mllib.IsotonicRegression, which used under the hood by ml.regression.IsotonicRegression. The previous implementation could have factorial complexity in the worst case. This implementation, which closely follows those in scikit-learn and the R `iso` package, runs in quadratic time in the worst case.
## How was this patch tested?

Existing unit tests in both `mllib` and `ml` passed before and after this patch. Scaling properties were tested by running the `poolAdjacentViolators` method in [scala-benchmarking-template](https://github.com/sirthias/scala-benchmarking-template) with the input generated by

``` scala
val x = (1 to length).toArray.map(_.toDouble)
val y = x.reverse.zipWithIndex.map{ case (yi, i) => if (i % 2 == 1) yi - 1.5 else yi}
val w = Array.fill(length)(1d)

val input: Array[(Double, Double, Double)] = (y zip x zip w) map{ case ((y, x), w) => (y, x, w)}
```

Before this patch:

| Input Length | Time (us) |
| --: | --: |
| 100 | 1.35 |
| 200 | 3.14 |
| 400 | 116.10 |
| 800 | 2134225.90 |

After this patch:

| Input Length | Time (us) |
| --: | --: |
| 100 | 1.25 |
| 200 | 2.53 |
| 400 | 5.86 |
| 800 | 10.55 |

Benchmarking was also performed with randomly-generated y values, with similar results.

### CODE REVIEW

**Comment by neggert** [comment]
Could someone please take a look at this? @mengxr maybe? Looks like you merged the initial implementation.

**Comment by neggert** [comment]
Anyone? @srowen?

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Nit: `j += 1` and likewise elsewhere.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: The condition changed to >= here on purpose right? you're now looking for a region that's non-increasing, not just strictly decreasing. Just checking.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: This condition is "if there was any decrease, not just == from i to j" right? and `pool` makes everything in [i,j] non-violating.

Comments from the previous code note that this could introduce a violation -- is this still possible? it wasn't obvious why it can't happen now.

**Review by srowen — CHANGES_REQUESTED** [review]
More than anything it would be great to add a couple simple unit tests of this method to prove it still works in key cases.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: That should be correct.  Lemma 1 from "A Fast Scaling Algorithm for Minimizing Separable Convex Functions Subject to Chain Constraints"

**Comment by jkbradley** [comment]
Thanks for pinging about this---it looks important!

One request: Do you know if there is a resource with a clearer statement of this precise algorithm for the L2 norm?  That'd be nice to have.  The best I found so far was "A Fast Scaling Algorithm for Minimizing Separable Convex Functions Subject to Chain Constraints"

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Is this comment correct? Looks like you pool the violating sequence [i, j] only. The preceding data points are checked for pooling in next outer loop.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: I think the original one i.e., `input(j)._1 > input(j + 1)._1` is correct. Here it is going to select out-of-order blocks.

Quoted from the paper:
> We refer to two blocks [p, q] and [q + 1, r] as consecutive. We refer to two consecutive blocks [p, q] and [q +1, r] as in-order if  theta_pq <= theta_q+1, r and out-of-order otherwise.

LEMMA 1 is pointing the how a merged block is also a single-valued block.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (neggert) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Correct. Any violations that are introduced will be fixed in the next iteration.

**Comment by neggert** [comment]
Given that the scikit-learn implementation I based this on has changed (https://github.com/scikit-learn/scikit-learn/pull/7444) since I submitted the PR, and @jkbradley has pointed out a reference I hadn't seen, I'd like to take a little bit of time to see if there's yet a better algorithm.

This one is an improvement over what we had, but as long as we're changing it, we may as well get it right.

**Comment by neggert** [comment]
Looks like the new scikit-learn implementation suffers from a similar problem to the one that the original Spark implementation had. I left them a note pointing it out.

Right now I'm working on implementing what's in "Minimizing Separable Convex Functions Subject to Simple Chain Constraints", by Best, Chakravarti, and Ubhaya. Running down the referecence trail, it looks like the original algorithm is in "Projections onto Order Simplexes" by Grotzinger and Witzgall, and this particular formulation of it is originally in "Active set algorithms for isotonic regression; A unifying framework" by Best and Chakravarti.

**Comment by srowen** [comment]
If it's correct and faster, I tend to favor merging this. The code looks OK to my understanding and passes existing tests.

**Comment by neggert** [comment]
Found another input that triggers non-polynomial time with the code in this PR. I'm again borrowing from scikit-learn. I think this is the case they found that led them to re-write their implementation.

```
    val y = ((0 until length) ++ (-(length - 1) until length) ++ (-(length - 1) to 0)).toArray.map(_.toDouble)
    val x = (1 to y.length).toArray.map(_.toDouble)
```

| Input Length | Time (ns) |
| --: | --: |
| 40 | 2059 |
| 80 | 4604 |
| 160 | 1974269 |
| 320 | 3246603433 |

I'm now working on implementing what's described in the Best papers. This should give O(n), even in the worst case. 

Should I close this and open a new PR with the new algorithm, or just add it here and you can squash when you merge?

**Comment by srowen** [comment]
Just add commits here.

**Comment by neggert** [comment]
Alright, the PAV algorithm has been completely re-written to follow what's outlined in "Minimizing Separable Convex Functions Subject to Simple Chain Constraints". I've tested it with a bunch of different inputs that caused previous version of this algorithm to go non-polynomial. It stays linear for all of them. I will note that it's slightly slower on very small datasets (< 5000 points or so), but those still finish in less than a millisecond on my laptop, so I'm not too concerned.

The only caveat is that there was one test that I couldn't get passing, so I removed it. It involved passing input data with 0 weights. I'd argue that the isotonic regression problem doesn't even have a unique solution in that case, so we shouldn't support it. Still, it is a slight change in behavior.

Looks like Jenkins is pointing out some style issues, so I'll get to work on those.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Super nit: space before `while` here and below. You can probably nix the extra parens around each of the two terms, but no big deal

**Review by srowen — COMMENTED** [review]
Modulo a few minor comments, the code at least looks clean. If the tests pass, seems OK.

**Comment by neggert** [comment]
@srowen Addressed your comments and fixed some style issues.

Some updated timings:

Alternating decreasing input

    val x = (1 to length).toArray.map(_.toDouble)
    val y = x.reverse.zipWithIndex.map{ case (yi, i) => if (i % 2 == 1) yi - 1.5 else yi}

Before: 

| Input Length | Time (us) |
|-------------:|----------:|
| 100 | 1.4 |
| 200 | 3.0 |
| 400 | 101.8 |
| 800 | 2,010,511.7 |

After:

| Input Length | Time (us) |
|-------------:|----------:|
| 100 | 2.5 |
| 200 | 5.8 |
| 400 | 10.1 |
| 800 | 20.7 |

Pathological input from scikit-learn benchmarks:

    val y= ((0 until length) ++ (-(length - 1) until length) ++ (-(length - 1) to 0)).toArray.map(_.toDouble)
    val x = (1 to y.length).toArray.map(_.toDouble)

Before:

| Input Length | Time (us) |
|-------------:|----------:|
| 40 | 2.2 |
| 80 | 4.7 |
| 160 | 2140.0 |
| 320 | 3,076,508.1 |

After:

| Input Length | Time (us) |
|-------------:|----------:|
| 40 | 5.3 |
| 80 | 10.4 |
| 160 | 21.3 |
| 320 | 41.9 |

**Comment by viirya** [comment]
For the zero-weight values, can we do similar to scikit-learn to remove zero-weight values, like https://github.com/amueller/scikit-learn/commit/2415100f79293bbbf52c12c36d63a6cf602cf3c4

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: I'm still not sure about this comment -- how can `blockBounds(x)` be both the start and end of a block? the implementation below is identical.

Comment (neggert) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: It relies on knowing ahead of time wether `x` is the start or end index<sup>1</sup>. If it's a start index, `blockBounds(x)` gives the ending index of that block. If it's an end index, `blockBounds(x)` gives the starting index of the block. So yes, the implementations of `blockStart` and `blockEnd` are identical. I just have two different functions because it makes the code more readable.

Maybe the comment from scikit-learn (where I borrowed this idea from) explains it better? (their `target` = my `blockBounds`)

> target describes a list of blocks.  At any time, if [i..j] (inclusive) is
> an active block, then blockBounds[i] := j and target[j] := i.

The trick is just in maintaining the array so that the above property is always true. At initialization, it's trivially true because all blocks have only one element, and `blockBounds(x)` = `x`. After initialization, `blockBounds` is only modified by the merge function, which is set up to modify `blockBounds` so that this property is preserved, then return the starting index of the newly-merged block.

This is admittedly a bit tricky, but it's a lot faster than the implementation I did where created a doubly-linked list of `Block` case classes. I'm open to suggestions on how to explain it better.

<sup>1</sup> You could actually figure out whether you have a start or an end index by comparing `blockBounds(x)` to `x`. The lesser value will be the start index.

**Comment by neggert** [comment]
@viirya Better to remove them, or throw an error? Personally, I'd rather be alerted that I'm passing invalid input, rather than have it "fixed" for me.

**Comment by viirya** [comment]
@neggert I am fine for throwing an error.

**Comment by srowen** [comment]
@neggert I think this is looking good except for the question about zero weights. If throwing an error is OK to you go for it. The test should be restored, but be updated to expect an exception. Other than that, more tests always help, but there is some OK coverage already

**Comment by srowen** [comment]
Ping @neggert

**Comment by neggert** [comment]
Sorry, I've been on vacation. I'll make the requested changes in the next day or two.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: I'd always write 0.0 instead of 0d for clarity. I also think we want an `IllegalArgumentException`, so maybe:

```
val weights = input.map { case (y, _, weight) =>
  require(weight > 0.0)
  (weight, weight * y)
}
```

**Review by srowen — CHANGES_REQUESTED** [review]
Just a minor style request

**Comment by neggert** [comment]
Done. It was surprisingly tricky to test, since the exception is thrown in the executor.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Are negative weights OK? (you don't need a type on `weight`, but hardly matters)

Comment (neggert) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: It's a weird thing to do, but I don't think there's any reason we can't support it. The problem still has a unique solution.

**Review by srowen — APPROVED** [review]
OK, seems reasonable to me. Any other thoughts @viirya ?

**Comment by viirya** [comment]
SGTM. See if @jkbradley has more comment?

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Please make the error message more informative:
* Make it clear that this indicates an internal bug within IsotonicRegression
* State the invalid values

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Please add an informative error message.

**Review by jkbradley — COMMENTED** [review]
Thanks for pinging me.  The updated implementation looks nice, but I do have comments about weights.

I'd like to disallow negative weights.  It's quite safe to assume that weights are non-negative in ML, and I've never seen negative weights used in isotonic regression.
* Q: I saw mention of negative weights above, but the Best and Chakravarti (1990) paper referenced explicitly requires strictly positive weights.  Does one of the other papers talk about negative weights?
* Btw, with negative weights, it would be easy to have undefined solutions.  Imagine the first element in the input having the smallest value and a negative weight.  A "correct" output would be infinite for that value and 0 elsewhere, giving a minimum objective value of NegativeInfinity.

I'd also prefer to allow zero weights to avoid breaking the API and since the model is fairly well-defined even with zero weights, except for edge cases at extreme feature values.  Removing the zero-weight rows should provide reasonable solutions, though we'd need to assert that the whole dataset had non-zero weight.

How does this sound?

**Comment by neggert** [comment]
Your changes sound fine. I don't have strong feelings one way or another, although I think we should at least throw a warning if we're discarding points. I do want to point out that disallowing negative weights also breaks the API, since they were previously supported (there's even a test for them).

**Review by srowen — APPROVED** [review]

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: Can you update the param doc to indicate that now negative weights are prohibited?

**Comment by viirya** [comment]
LGTM

**Comment by jkbradley** [comment]
@neggert I'll ask @mengxr about the negative weights since he oversaw the original work here.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: I disagree about logging for zero-weight points.  If an instance has zero weight, then it should be ignored, and that's not a problem.

If we really want to log this, then let's do it once per dataset, not once per row, which could make logs explode.  I'd also say we should lower the logging level to logInfo.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: State invalid y,x,weight.

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: filter would be simpler than flatMap

File: mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala]: missing "]"

**Review by jkbradley — COMMENTED** [review]
Just a few comments, other than checking with @mengxr about negative weights.

**Comment by jkbradley** [comment]
@zapletal-martin  Pinging since you wrote the original PR: There's discussion here about whether IsotonicRegression should support negative weights.  Is there a good reason to?  I haven't seen papers which allow negative weights, though I've only looked at a few.  Thanks!

**Comment by neggert** [comment]
My thought was that there's nothing in the computation that prevents it (unlike 0 weights, which cause NaNs). Why not err on the side of maximum flexibility to the end user? Just because there's no use case for it now doesn't mean someone won't find one in the future.

Just my 2 cents. I don't feel strongly about it and will defer to commiters.

**Comment by neggert** [comment]
Looks like the build completed successfully. Somehow Jenkins didn't post the message.

**Comment by jkbradley** [comment]
Thanks for the updates!  This LGTM, except for deciding about negative weights.

Responding to your comment above, negative weights are just as problematic as 0 weights.  See my comment above "Btw, with negative weights, ..."

I spoke with @mengxr offline, and he does not recall why the original work supported negative weights.  I'll wait a few days before merging this to see if @zapletal-martin can check out the issue.

**Comment by srowen** [comment]
@jkbradley I'll merge if there are no other thoughts here. Seems OK to me too

**Comment by jkbradley** [comment]
Sounds good.  I'll run fresh tests before merging to be safe though.

**Comment by jkbradley** [comment]
Merging with master
Thanks!

**Pull request closed without merging**
